File: tests/test_feedback.cpp
/**
 * test_feedback.cpp — Unit tests for the feedback engine.
 *
 * Uses a lightweight custom test framework (no external dependencies).
 * Tests all edge cases, especially duplicate-letter handling.
 *
 * Build & run:
 *   make test
 *   # or: g++ -std=c++17 -o test_feedback tests/test_feedback.cpp src/feedback.cpp && ./test_feedback
 */

#include "../src/feedback.h"
#include <iostream>
#include <string>
#include <vector>

// -----------------------------------------------------------------------
// Minimal test framework
// -----------------------------------------------------------------------
static int total = 0, passed = 0, failed_count = 0;

#define EXPECT_EQ(name, got, expected)                                          \
    do {                                                                        \
        total++;                                                                \
        if ((got) == (expected)) {                                              \
            passed++;                                                           \
            std::cout << "  PASS  " << (name) << "\n";                         \
        } else {                                                                \
            failed_count++;                                                     \
            std::cout << "  FAIL  " << (name) << "\n"                          \
                      << "        got:      \"" << (got) << "\"\n"              \
                      << "        expected: \"" << (expected) << "\"\n";        \
        }                                                                       \
    } while (0)

void section(const std::string& name) {
    std::cout << "\n[" << name << "]\n";
}

// -----------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------
void test_basic() {
    section("Basic Cases");

    // Perfect match — all green
    EXPECT_EQ("perfect match 'hello'", get_feedback("hello", "hello"), "22222");
    EXPECT_EQ("perfect match 'crane'", get_feedback("crane", "crane"), "22222");

    // All gray — no letters in common
    EXPECT_EQ("all gray 'abcde'→'fghij'", get_feedback("abcde", "fghij"), "00000");
    EXPECT_EQ("all gray 'zzzzz'→'aaaaa'", get_feedback("zzzzz", "aaaaa"), "00000");

    // All yellow — right letters, all wrong positions
    // abcde vs bcdea: each letter shifted by one position
    EXPECT_EQ("all yellow 'abcde'→'bcdea'", get_feedback("abcde", "bcdea"), "11111");
}

void test_mixed() {
    section("Mixed Green/Yellow/Gray");

    // crane vs trace: t(0)r(1)a(2)c(3)e(4)
    // c(0)!=t → pool check: c in [t,c] (pos 3 unmatched) → yellow
    // r(1)==r → green
    // a(2)==a → green
    // n(3)!=c → pool check: n not in [t] → gray
    // e(4)==e → green
    // Result: 1 2 2 0 2
    EXPECT_EQ("crane→trace", get_feedback("crane", "trace"), "12202");

    // speed vs creep: c(0)r(1)e(2)e(3)p(4)
    // s(0)!=c → gray
    // p(1)!=r → pool check: p in [c,r,p]? yes → yellow (pool[4] consumed)
    // e(2)==e → green
    // e(3)==e → green
    // d(4)!=p → pool=[c,r] → gray
    // Result: 0 1 2 2 0
    EXPECT_EQ("speed→creep", get_feedback("speed", "creep"), "01220");
}

void test_duplicate_in_guess() {
    section("Duplicate Letters in Guess");

    // guess has 'a' twice, answer has 'a' twice — both green
    // aabbb vs aaccc: a(0)==a→green, a(1)==a→green, b's not in pool [c,c,c] → gray
    EXPECT_EQ("aabbb→aaccc (aa both green)", get_feedback("aabbb", "aaccc"), "22000");

    // aaaaa vs aabbb: first two a's are green, rest gray (no more a's in pool)
    EXPECT_EQ("aaaaa→aabbb", get_feedback("aaaaa", "aabbb"), "22000");

    // speed vs speed: perfect
    EXPECT_EQ("speed→speed (perfect)", get_feedback("speed", "speed"), "22222");
}

void test_duplicate_in_answer() {
    section("Duplicate Letters in Answer");

    // crane vs creep: c(0)r(1)e(2)e(3)p(4)
    // c(0)==c→green, r(1)==r→green
    // a(2): pool=[e,e]→gray, n(3): pool=[e,e]→gray, e(4): e in pool→yellow
    EXPECT_EQ("crane→creep", get_feedback("crane", "creep"), "22001");

    // eerie vs geese: g(0)e(1)e(2)s(3)e(4)
    // e(0)!=g → pass2: e in [g,e,s]→yellow (consume pos1 e)
    // e(1)==e→green (consume)
    // r(2)!=e → r not in remaining [g,s]→gray
    // i(3)!=s → i not in [g,s]→gray
    // e(4)==e→green
    EXPECT_EQ("eerie→geese", get_feedback("eerie", "geese"), "12002");
}

void test_yellow_cap() {
    section("Yellow Count Capped by Answer Occurrences");

    // aacde vs bbaaa: b(0)b(1)a(2)a(3)a(4)
    // a(0): pool=[b,b,a,a,a]→yellow (consume one a)
    // a(1): pool=[b,b,a,a]→yellow (consume one a)
    // c(2): c not in pool→gray
    // d(3): d not in pool→gray
    // e(4): e not in pool→gray
    EXPECT_EQ("aacde→bbaaa (yellows capped)", get_feedback("aacde", "bbaaa"), "11000");

    // guess all same, answer has one: aaaaa vs bbbba
    // a(0)!=b, a(1)!=b, a(2)!=b, a(3)!=b, a(4)==a→green
    // Pool after pass1: [b,b,b,b]
    // Pass2: none of a(0-3) in [b,b,b,b]→all gray
    EXPECT_EQ("aaaaa→bbbba", get_feedback("aaaaa", "bbbba"), "00002");
}

void test_filter() {
    section("Filter Candidates");

    std::vector<std::string> candidates = {"crane", "trace", "plant", "chess"};

    // After guessing "crane" and getting all-green pattern, only "crane" remains
    auto r1 = filter_candidates(candidates, "crane", "22222");
    EXPECT_EQ("filter perfect match", r1.size() == 1 && r1[0] == "crane", true);

    // After getting all-gray, crane should be removed
    auto r2 = filter_candidates(candidates, "crane", "00000");
    bool no_crane = true;
    for (const auto& w : r2) if (w == "crane") no_crane = false;
    EXPECT_EQ("filter removes crane on 00000", no_crane, true);
}

// -----------------------------------------------------------------------
// Main
// -----------------------------------------------------------------------
int main() {
    std::cout << "=========================================\n";
    std::cout << "  Wordle Solver — Feedback Engine Tests\n";
    std::cout << "=========================================\n";

    test_basic();
    test_mixed();
    test_duplicate_in_guess();
    test_duplicate_in_answer();
    test_yellow_cap();
    test_filter();

    std::cout << "\n=========================================\n";
    std::cout << "  Results: " << passed << "/" << total << " passed";
    if (failed_count == 0) {
        std::cout << "  ALL TESTS PASSED ✓\n";
    } else {
        std::cout << "  " << failed_count << " FAILED\n";
    }
    std::cout << "=========================================\n";

    return (failed_count == 0) ? 0 : 1;
}

File: tests/test_feedback.py
"""
test_feedback.py — Unit tests for the feedback function.

Run with: pytest tests/test_feedback.py -v

These tests are critical — the entire solver's correctness depends on
the feedback function being exactly right, especially for duplicate letters.
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

import pytest
from feedback import get_feedback, filter_candidates, pattern_to_emoji


class TestBasicFeedback:
    """Basic feedback function correctness tests."""

    def test_perfect_match(self):
        """All greens when guess == answer."""
        assert get_feedback("crane", "crane") == "22222"
        assert get_feedback("hello", "hello") == "22222"
        assert get_feedback("abcde", "abcde") == "22222"

    def test_all_gray(self):
        """All grays when no letters match."""
        assert get_feedback("abcde", "fghij") == "00000"
        assert get_feedback("zzzzz", "aaaaa") == "00000"

    def test_all_yellow(self):
        """All yellows — right letters, all wrong positions."""
        # "abcde" vs "bcdea" — each letter is present but shifted
        result = get_feedback("abcde", "bcdea")
        assert result == "11111"

    def test_mixed_simple(self):
        """Simple mixed case with no duplicates."""
        # crane vs trace: t-r-a-c-e
        # Pass1 greens: r(1)==r, a(2)==a, e(4)==e  →  pool remaining: [t, c]
        # Pass2: c(0) → yellow (c in [t,c]), n(3) → gray
        # Result: c=yellow(1), r=green(2), a=green(2), n=gray(0), e=green(2) = "12202"
        result = get_feedback("crane", "trace")
        assert result == "12202"


class TestDuplicateLetters:
    """
    Duplicate letter handling is the trickiest part of the feedback function.
    These tests specifically target edge cases.
    """

    def test_duplicate_in_guess_not_answer(self):
        """Guess has a duplicate letter that the answer only has once."""
        # guess="aabbb", answer="aaccc"
        # a(0): a(0) → GREEN. a(1): a(1) → GREEN. b(2),(3),(4): not in remaining → GRAY
        # Pool after greens: [c, c, c] (positions 2,3,4)
        # Pass 2: b's are not in [c,c,c] → all gray
        assert get_feedback("aabbb", "aaccc") == "22000"

    def test_duplicate_in_answer_not_guess(self):
        """Answer has a duplicate letter but guess only has it once."""
        # guess="crane", answer="creep"
        # creep: c(0),r(1),e(2),e(3),p(4)
        # crane: c(0),r(1),a(2),n(3),e(4)
        # Pass1: c(0)==c→GREEN, r(1)==r→GREEN, a(2)!=e, n(3)!=e, e(4)!=p
        # Pool after: [e, e] (positions 2,3 unmatched)
        # Pass2: a(2): not in [e,e] → GRAY. n(3): not in [e,e] → GRAY.
        # e(4): e in [e,e] → YELLOW
        # Result: 2,2,0,0,1 = "22001"
        assert get_feedback("crane", "creep") == "22001"

    def test_guess_repeats_letter_seen_green(self):
        """Guess has same letter at two positions; answer has it twice (both green!)."""
        # guess="speed", answer="creep"
        # creep: c(0),r(1),e(2),e(3),p(4)
        # speed: s(0),p(1),e(2),e(3),d(4)
        # Pass1: e(2)==e→GREEN, e(3)==e→GREEN
        # Pool after: [c, r, p]
        # Pass2: s(0): not in [c,r,p] → GRAY. p(1): p in [c,r,p] → YELLOW. d(4): not → GRAY
        # Result: 0,1,2,2,0 = "01220"
        assert get_feedback("speed", "creep") == "01220"

    def test_all_same_letter_in_guess(self):
        """Guess is all same letter."""
        # guess="aaaaa", answer="aabbb"
        # a(0)==a(0) → GREEN. a(1)==a(1)... wait answer[1]=a? "aabbb": a(0),a(1),b(2),b(3),b(4)
        # a(0)==a(0) → GREEN
        # a(1)==a(1) → GREEN
        # a(2)!=b → no green
        # a(3)!=b → no green
        # a(4)!=b → no green
        # Pool after: [b, b, b] (positions 2,3,4)
        # Pass 2: none of the remaining a's are in [b,b,b] → all GRAY
        assert get_feedback("aaaaa", "aabbb") == "22000"

    def test_yellow_only_counted_once(self):
        """Same letter should only get yellow once if answer has it once."""
        # guess="aacde", answer="bbaaa"
        # answer: b(0),b(1),a(2),a(3),a(4)
        # a(0)!=b → no green
        # a(1)!=b → no green
        # c(2)!=a → no green
        # d(3)!=a → no green
        # e(4)!=a → no green
        # Pool: [b,b,a,a,a]
        # Pass 2:
        # a(0): a in pool → YELLOW, consume one a → pool=[b,b,a,a]
        # a(1): a in pool → YELLOW, consume one a → pool=[b,b,a]
        # c(2): c not in pool → GRAY
        # d(3): d not in pool → GRAY
        # e(4): e not in pool → GRAY
        assert get_feedback("aacde", "bbaaa") == "11000"

    def test_multiple_yellows_limited_by_answer(self):
        """Multiple yellows only up to count in answer."""
        # guess="eerie", answer="geese"
        # answer: g(0),e(1),e(2),s(3),e(4)
        # e(0)!=g → no green
        # e(1)==e(1) → GREEN
        # r(2)!=e → no green
        # i(3)!=s → no green
        # e(4)==e(4) → GREEN
        # Pool after: [g, e, s] (positions 0,2,3 unmatched)
        # Pass 2:
        # e(0): e in [g,e,s] → YELLOW, consume e → pool=[g,s]
        # r(2): r not in [g,s] → GRAY
        # i(3): i not in [g,s] → GRAY
        # Result: e=1,e=2,r=0,i=0,e=2 = "12002"
        assert get_feedback("eerie", "geese") == "12002"


class TestFilterCandidates:
    """Test the candidate filtering function."""

    def test_filter_removes_inconsistent(self):
        """Filter should keep only words consistent with the feedback."""
        candidates = ["crane", "trace", "raise", "arise", "saner", "lanes"]
        # Guess "crane" with pattern "12201" should keep words that produce that
        # pattern when "crane" is guessed against them
        guess = "crane"
        pattern = get_feedback("crane", "trace")  # whatever feedback crane→trace gives
        filtered = filter_candidates(candidates, guess, pattern)
        assert "trace" in filtered

    def test_filter_perfect_match(self):
        """With pattern 22222, only the exact answer remains."""
        candidates = ["crane", "trace", "plant"]
        filtered = filter_candidates(candidates, "crane", "22222")
        assert filtered == ["crane"]

    def test_filter_all_gray(self):
        """All gray means none of those letters appear."""
        # "abcde" vs any word with no a,b,c,d,e
        candidates = ["fghij", "abcde", "xxxxx", "aaaaa"]
        filtered = filter_candidates(candidates, "abcde", "00000")
        # Only words with none of a,b,c,d,e should remain
        assert "fghij" in filtered or "xxxxx" in filtered
        assert "abcde" not in filtered  # Would give 22222, not 00000


class TestPatternToEmoji:
    """Test the emoji display utility."""

    def test_all_green(self):
        assert pattern_to_emoji("22222") == "🟩🟩🟩🟩🟩"

    def test_all_yellow(self):
        assert pattern_to_emoji("11111") == "🟨🟨🟨🟨🟨"

    def test_all_gray(self):
        assert pattern_to_emoji("00000") == "⬜⬜⬜⬜⬜"

    def test_mixed(self):
        assert pattern_to_emoji("21010") == "🟩🟨⬜🟨⬜"


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
